Add data interface for Holographic Stimulation data

alessandratrapani opened the pull request:
The holographic stimulation data are just for the "5stim" and "7expt" sessions
From the README provided in the data folder:
**Variables and nomenclature**
`hologram_list` = list of unique stimulations/holograms that is indexed by `stim_id`
`targeted_cells` = overally list of all suite2p cells targeted and matched to holographic stimuli
`stim_times` = stimulation time (in seconds) within a trial for each roi, NaN means the cell was not stimulated
`stim_id` = trialwise index into `hologram_list`, note this is idx-1, as the first "stim" condition is a control trial
`power_per_cell` = power per cell/stimulation site in mW
`spikes_per_cell` = stim frequency per cell/stimulation site _<span style="color: red;">what is the difference from hz_per_cell?</span>.-> from email: it's a typo, it actually means the number of stimulation per cell/stimulation site_
`hz_per_cell` = stimulationd frequency per cell/stimulation site
`frame_rate` = aquisition frame rate in Hz _<span style="color: red;">missing</span> -> from email: it's the same as in Suite2P_

##### Session "5stim"
In "5stim" there are 24 trials. Each trial correspond to a multipage .tif file in the "5stim" imaging folder. There are 36 combinations of max 10 ROIs (over a total of 368-->`len(targeted_cells)`) stimulated at the same time (`hologram_list.shape: (36,10)`). Only the first 8 combinations are referred in `stim_id`, the actual values of `stim_id` range from 0 (that is a control trial where there is no actual stimulation: `power_per_cell` and `spikes_per_cell` are 0) to 8 (`hologram_list[7]`).

##### Session "7expt"
In "7expt" there are 723 trials (`len(stim_id): 723`). Each trial correspond to a multipage .tif file in the "7expt" imaging folder, only 209 .tif files in the folder--> some files are missing due to problem during transfer. 10 ROIs stimulated in separate trials alone (`hologram_list.shape: (10,1)`). All ROIs in hologram list are stimulated at least once: `stim_id` range from 0 -control trial- to 10 (`hologram_list[9]`).
NB: for `stim_id`=4 --> `hologram_list[3]` is nan <span style="color: red;">why?</span> 

#### ROIs indices
In the example.hdf5 file the targeted cells are 368 (`len(targeted_cells)`) in total, both for 5stim and 7expt. `targeted_cells` value span from 0 to 700, that are the ids of the accepted ROIs segmented with Suite2p concatenated over the three planes:
(`len(iscell_0[iscell_0[:,0]!=0.])+len(iscell_1[iscell_1[:,0]!=0.])+len(iscell_2[iscell_2[:,0]!=0.])`). 

In the ScanImage .tif metadata there is a dictonary saved as text that the function `extract_extra_metadata` cannot parse, where information on the ROIs are stored. I implemented a new function call `extract_roi_metadata` (https://github.com/catalystneuro/roiextractors/pull/272) that return this dict.
```
"RoiGroups":{
    "imagingRoiGroup":{# general imaging space metadata (only 2D)
        ...
        "centerXY": [-1.776356839e-16,0.1],
        "sizeXY": [13.5,13.5], # size of the FOV in micrometers
        "rotationDegrees": 0,
        "enable": 1,
        "pixelResolutionXY": [512,512],
        "pixelToRefTransform": [[0.0263671875,0,-6.763183594],
                                [0,0.263671875,-6.663183594],
                                [0,0,1]],
        "affine":  [[13.5,0,-6.75],
                    [0,13.5,-6.65],
                    [0,0,1]]
        ...
    },
    "integrationRoiGroup":{
        ...
        "rois": [
        {
          "ver": 1,
          "classname": "scanimage.mroi.Roi",
          "name": "ROI 1",
            ...
          "zs": 0,
          "scanfields": {
            ...
            "centerXY": [2.518066406,1.062402344],
            "sizeXY": [0.2373046875,0.2373046875],
            "rotationDegrees": 0,
            "enable": 1,
            "threshold": 100,
            "channel": 1,
            "processor": "cpu",
            "mask": [roi_pixel_mask 9x9 binary matrix],
            "affine": [[0.2373046875,0,2.518066406],
                       [0,0.2373046875,1.062402344],
                       [0,0,1]]
          },
          "discretePlaneMode": 1,
          "powers": null,
          "pzAdjust": null,
          "Lzs": null,
          "interlaceDecimation": null,
          "interlaceOffset": null,
          "enable": 1
        },
        ... # all the 368 ROIs
        ]
    } # imaging space metadata for each ROI
}
```
From `imagingRoiGroup` we can extract info on the dimension of the field of view (XY) from `"sizeXY"` and divide by the dimension to get the grid_spacing (z dimension missed)

In `integrationRoiGroup` we can extract information about the position of the targeted rois: inv(`pixelToRefTransform`)*`centerXY`

alessandratrapani commented:
I still have some issues with the extension to make segmented_rois not required. I am working on this

weiglszonja commented on src/mousev1_to_nwb/abdeladim_2023/abdeladim_2023holostiminterface.py:
```diff
@@ -0,0 +1,242 @@
+
+        Parameters
+        ----------
+        folder_path : FolderPathType
+            The folder path that contains the holographic stimulation data.
```
[minor] I think it would make sense to clarify this the folder with the imaging data as well, otherwise it is confusing how different it is from `holographic_stimulation_file_path` which does not have a docstring.

alessandratrapani replied:
Good catch, thanks! it's updated now

weiglszonja commented on src/mousev1_to_nwb/abdeladim_2023/abdeladim_2023holostiminterface.py:
```diff
@@ -0,0 +1,242 @@
+
+        # Add SLM device to NWBFile
+        spatial_light_modulator_metadata = metadata_copy["Ophys"]["OptogeneticDevice"]["SpatialLightModulator2D"]
+        spatial_light_modulator_metadata.update(spatial_resolution=[1024, 1024])
```
I think it is better to do this in `interface.get_metadata()`

weiglszonja commented on src/mousev1_to_nwb/abdeladim_2023/abdeladim_2023holostiminterface.py:
```diff
@@ -0,0 +1,242 @@
+        fov_size_in_um = np.array(self.rois_metadata["imagingRoiGroup"]["rois"]["scanfields"]["sizeXY"])
+        frame_dimesion = np.array(self.rois_metadata["imagingRoiGroup"]["rois"]["scanfields"]["pixelResolutionXY"])
+
+        imaging_plane_name = "ImagingPlaneHolographicStimulation"
+        imaging_plane = nwbfile.create_imaging_plane(
+            name=imaging_plane_name,
+            optical_channel=optical_channel,
+            imaging_rate=float(
+                self.image_metadata["SI.hRoiManager.scanFrameRate"]
+            ),  # Frame rate on the entire volumetric stack
+            description="Imaging plane for the holographic stimulation",
+            device=device,
+            excitation_lambda=600.0,  # TODO update
+            indicator="GFP",  # TODO update
+            location="V1",  # TODO update
+            grid_spacing=fov_size_in_um / frame_dimesion,
+            grid_spacing_unit="micrometers",
```
Is there a reason why not use the utility method `add_imaging_plane() `from neuroconv for this?

alessandratrapani replied:
If I do it like this:
```python
        imaging_plane_name = "ImagingPlaneHolographicStimulation"
        imaging_plane_metadata = metadata_copy["Ophys"]["ImagingPlane"][-1]
        imaging_plane_metadata.update(
            name=imaging_plane_name,
            optical_channel=optical_channel,
            imaging_rate=float(
                self.image_metadata["SI.hRoiManager.scanFrameRate"]
            ),  # Frame rate on the entire volumetric stack
            description="Imaging plane for the holographic stimulation.",
            device=device,
            excitation_lambda=920.0,
            indicator="GCaMP6s", 
            location="Primary visual cortex (V1), 140-200 um below pia", 
            grid_spacing=fov_size_in_um / frame_dimesion,
            grid_spacing_unit="micrometers",
            origin_coords=self.rois_metadata["imagingRoiGroup"]["rois"]["scanfields"]["centerXY"],
            origin_coords_unit="micrometers",
        )      

add_imaging_plane(nwbfile=nwbfile,metadata=metadata_copy,imaging_plane_name=imaging_plane_name)

```
I got this error:
`The `._to_dict()` method for pickling has not been defined for this DataChunkIterator!`

weiglszonja replied:
yeah just make sure you're not passing objects to metadata `dict`, e.g. for optical channel if this is a metadata update then it should be the name of the OpticalChannel, same for device which should be the name of the Device.

weiglszonja commented on src/mousev1_to_nwb/abdeladim_2023/abdeladim_2023holostiminterface.py:
```diff
@@ -0,0 +1,242 @@
+                    targeted_rois = targeted_plane_segmentation.create_roi_table_region(
+                        name="targeted_rois",
+                        description="targeted rois",
+                        region=[
+                            i
+                            for i, roi_index in enumerate(self._targeted_to_segmented_roi_ids_map)
+                            if roi_index in roi_indexes
+                        ],
```
I think this part would be more readable if you assign a variable to it first, but keep it like this if you prefer it this way

alessandratrapani replied:
thanks! I changed that

weiglszonja commented:
Overall LGTM, can you share an example stubbed file when you feel ready?

alessandratrapani commented:
> Overall LGTM, can you share an example stubbed file when you feel ready?

yes! I was waiting for it to upload: https://catalystneuro-processing.s3.us-east-2.amazonaws.com/adesnik-lab-to-nwb/2023119_w57_1_5stim_test.nwb

weiglszonja commented:
> > Overall LGTM, can you share an example stubbed file when you feel ready?
> 
> yes! I was waiting for it to upload: https://catalystneuro-processing.s3.us-east-2.amazonaws.com/adesnik-lab-to-nwb/2023119_w57_1_5stim_test.nwb

I noticed there are two microscope devices added to the file, the one with the default name "Microscope" and "CustomMicroscope", where `"CustomMicroscope"` is linked to `"ImagingPlaneHolographicStimulation"` and the default named is linked to the other imaging planes. So the imaging and stimulation was carried out with different microscopes then? 

Also it is a bit weird, but I'm not that familiar with this dataset; why there are two optical channels assigned to each imaging planes? E.g. "Chan1" and "Green", is this intentional or the metadata was not updated somewhere and we also added the default optical channel?

alessandratrapani commented:
> I noticed there are two microscope devices added to the file, the one with the default name "Microscope" and "CustomMicroscope", where `"CustomMicroscope"` is linked to `"ImagingPlaneHolographicStimulation"` and the default named is linked to the other imaging planes. So the imaging and stimulation was carried out with different microscopes then?

No, it should be just `"CustomMicroscope"`, but the suite2p interface create "Microscope" and I am having some trouble understanding how to remove it (#8). 

> Also it is a bit weird, but I'm not that familiar with this dataset; why there are two optical channels assigned to each imaging planes? E.g. "Chan1" and "Green", is this intentional or the metadata was not updated somewhere and we also added the default optical channel?

Sorry for the confusion, it was not updated in the segmentation interface and added the default optical channel. Should be solved now, here: #8 but I am not sure this is the right way.

weiglszonja commented:
> > I noticed there are two microscope devices added to the file, the one with the default name "Microscope" and "CustomMicroscope", where `"CustomMicroscope"` is linked to `"ImagingPlaneHolographicStimulation"` and the default named is linked to the other imaging planes. So the imaging and stimulation was carried out with different microscopes then?
> 
> No, it should be just `"CustomMicroscope"`, but the suite2p interface create "Microscope" and I am having some trouble understanding how to remove it (#8).

Have you tried overriding it in the converter's get_metadata() function? [This](https://github.com/catalystneuro/pinto-lab-to-nwb/blob/6ca518c15d34dd7bc7786ec8807696c482d43842/src/pinto_lab_to_nwb/into_the_void/into_the_voidnwbconverter.py#L161) might help you with that.

weiglszonja commented:
@alessandratrapani let me know if this is ready for final review :)

alessandratrapani commented:
@weiglszonja should be ready. The file for this session example should be uploaded in a bit.

weiglszonja reviewed (approved):
LGTM, checked the file, ophys metadata looks great!
